persist/sqldb/workflow_archive.go

```diff
@@ -178,48 +187,35 @@ func (r *workflowArchive) ListWorkflows(options sutils.ListOptions) (wfv1.Workfl
 
 	wfs := make(wfv1.Workflows, len(archivedWfs))
 	for i, md := range archivedWfs {
-		labels := make(map[string]string)
-		if err := json.Unmarshal([]byte(md.Labels), &labels); err != nil {
-			return nil, err
-		}
-		// For backward compatibility, we should label workflow retrieved from DB as Persisted.
-		labels[common.LabelKeyWorkflowArchivingStatus] = "Persisted"
-
-		annotations := make(map[string]string)
-		if err := json.Unmarshal([]byte(md.Annotations), &annotations); err != nil {
-			return nil, err
-		}
-
-		t, err := time.Parse(time.RFC3339, md.CreationTimestamp)
+		var wf workflow
+		// NOTE: Switched from SQL JSON extraction to Golang JSON parsing for better performance. #13382
```

Comment:
Did you test this? It would be good to see metrics before/after.

It was actually done in Go first and then switched to SQL because the Go variant required the entire data structure to be retrieved instead of the pieces needed (and the entire `status` can be very large) which was far too slow and unnecessary.

This is also not the solution I suggested in #13382 nor #13295

Reply:
> Did you test this? Your "Verification" section is empty.

I have tested the workflow submit and list with a small data, but not yet tested it with large data.

> Per #12912

I have reduced the size of the Go struct for `json.Unmarshal`. Benchmark results show that resource usage has decreased compared to #12912. However, while it not be directly compared to SQL, this method might still be insufficient for improvement.

Benchmark results
```
Benchmark_Unmarshal1
Benchmark_Unmarshal1-10    	   15165	     76526 ns/op	 359.04 MB/s	   70060 B/op	    1230 allocs/op
Benchmark_Unmarshal2
Benchmark_Unmarshal2-10    	   45692	     27174 ns/op	1011.11 MB/s	    4720 B/op	      81 allocs/op
PASS
```

Here is the Benchmark code
```go
package sqldb

import (
	"encoding/json"
	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
	"io"
	"os"
	"testing"
)

var codeJSON []byte

func codeInit() {
	f, err := os.Open("test.json")
	if err != nil {
		panic(err)
	}
	defer f.Close()
	data, err := io.ReadAll(f)
	if err != nil {
		panic(err)
	}

	codeJSON = data
}

func Benchmark_Unmarshal1(b *testing.B) {
	b.ReportAllocs()
	if codeJSON == nil {
		b.StopTimer()
		codeInit()
		b.StartTimer()
	}
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			var r wfv1.Workflow
			if err := json.Unmarshal(codeJSON, &r); err != nil {
				b.Fatalf("Unmarshal error: %v", err)
			}
		}
	})
	b.SetBytes(int64(len(codeJSON)))
}

func Benchmark_Unmarshal2(b *testing.B) {
	b.ReportAllocs()
	if codeJSON == nil {
		b.StopTimer()
		codeInit()
		b.StartTimer()
	}
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			var r workflow
			if err := json.Unmarshal(codeJSON, &r); err != nil {
				b.Fatalf("Unmarshal error: %v", err)
			}
		}
	})
	b.SetBytes(int64(len(codeJSON)))
}

```

Reply:
> I have reduced the size of the Go struct for `json.Unmarshal`

Yes but this method still transmits the entirety of the data over the network and still has to be parsed, even if not allocated.

For the Controller estimation scenario of your issue, I had suggested removing the unnecessary columns from the query entirely. It can use a different but similar query.

For the Server API, I had suggested a materialized view and others suggested JSONB (only Postgres compatible, so not a complete solution) and other derivations.

I think all of those methods are better than this one

Reply:
> For the Controller estimation scenario of your issue, I https://github.com/argoproj/argo-workflows/issues/13382#issuecomment-2245467582 removing the unnecessary columns from the query entirely. It can use a different but similar query.

I think this approach is good. I will revert this Pull Request to Draft.

Reply:
I have updated this Pull Request, rebased the previous changes and force pushed.